multi-label classification example: CelebA image tagging

aditya1503 opened the pull request:
Adds model training code for CelebA to accompany cleanlab's main multi-label classification tutorial.

aditya1503 committed:
adding tutorial

aditya1503 committed:
val loss fix

aditya1503 committed:
load pred probs

aditya1503 committed:
tutorial notebook

cgnorthcutt commented:
Remember to add a 3 lines of code snippet at the top if possible that highlights example of the main methods and how to use them (doesn't need to be runnable) -- in a markdown block.

All tutorials and docstrings should have this eventually.

aditya1503 committed:
k fold

jwmueller commented on multi_label_classification/training.ipynb:
```diff
@@ -0,0 +1,739 @@
+{
```
Let's call this folder: 
`multilabel_classification/`

jwmueller replied:
Let's call this notebook: 
`pytorch_network_training.ipynb`

jwmueller commented on multi_label_classification/tutorial-predprobs.ipynb:
```diff
@@ -0,0 +1,249 @@
+{
```
Move this to main cleanlab repo. There let's name this file:
`multilabel_classification.ipynb`

jwmueller replied:
The tutorial should start with a markdown cell with the name of the tutorial:

\# Find label issues in Multi-Label Classification Data

aditya1503 replied:
Added to first cell

aditya1503 committed:
training notebook cleanup;

aditya1503 committed:
moving

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import timm\n",
+    "from torch import nn\n",
+    "from timm.data.loader import create_loader\n",
+    "import torch.utils.data as data\n",
+    "import torch\n",
+    "import pandas as pd\n",
+    "\n",
+    "from sklearn.metrics import balanced_accuracy_score\n",
+    "\n",
+    "\n",
+    "from imp import reload\n",
+    "from collections import defaultdict\n",
+    "from types import SimpleNamespace\n",
+    "\n",
+    "\n",
+    "import numpy as np\n",
+    "from sklearn.metrics import accuracy_score\n",
+    "from collections import Counter\n",
+    "from timm.optim import create_optimizer\n",
+    "import pickle"
```
```suggestion
    "import pickle",
    "from imp import reload\n",
    "from collections import defaultdict, Counter\n",
    "from types import SimpleNamespace\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.metrics import balanced_accuracy_score\n",
    "from sklearn.metrics import accuracy_score\n",
    "import torch\n",
    "from torch import nn\n",
    "import torch.utils.data as data\n",
    "import timm\n",
    "from timm.data.loader import create_loader\n",
    "from timm.optim import create_optimizer"
```

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from timm.data import (\n",
+    "    resolve_data_config,\n",
+    ")\n",
+    "from timm.models import create_model"
```
let's move all imports to top of notebook

aditya1503 replied:
done : https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df = pd.read_csv(\"img_align_celeba/all.csv\")\n",
```
need to get this file from somewhere?  This example should be programmatically runnable

aditya1503 replied:
Yes the wget scripts will be added here,

aditya1503 replied:
https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.metrics import balanced_accuracy_score\n",
+    "from sklearn.metrics import accuracy_score"
```
unnecessary

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+{
```
First cell should always be markdown with a title and short description of what the notebook does.

In this case, title can be:  Training a neural network for multi-label classification on the Celeb-A dataset

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   },
+   "outputs": [
+    {
+     "name": "stdout",
```
Most people will view this notebook through github's viewer:
https://github.com/cleanlab/examples/blob/fbb88a4055cb7f942520dc59409551a559de71dc/multilabel_classification/pytorch_network_training.ipynb

You can see this output cell is huge and takes forever to scroll through, so it would be good to print less output by default (you can add `verbosity` arg to your model for example and reduce the verbosity level to suppress constant printing of the loss)

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "<module 'os' from '/usr/lib/python3.8/os.py'>"
+      ]
+     },
+     "execution_count": 20,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
```
get rid of this?  not sure what it is doing here

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+    "pred_probs = df_pred[cols].to_numpy()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 32,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "np.save(open('pred_probs.npy','wb'),pred_probs)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 36,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "data = list(df_pred['image_loc'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 38,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "<module 'pickle' from '/usr/lib/python3.8/pickle.py'>"
+      ]
+     },
+     "execution_count": 38,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "pickle.dump(data,open(\"data.p\",'wb'))"
+   ]
```
these can all be one cell that is used to save all the results.

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  }
```
delete this empty cell

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
+}
```
Don't worry about polishing the text too much, I'll do a rewrite of the text for you to make the notebook as clear as possible

aditya1503 committed:
Update multilabel_classification/pytorch_network_training.ipynb

Co-authored-by: Jonas Mueller <1390638+jwmueller@users.noreply.github.com>

jwmueller commented on multilabel_classification/pytorch_network_training.ipynb:
```diff
@@ -0,0 +1,1064 @@
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
```
You'll want to add a requirements.txt file in this folder that lists the dependencies (and specific versions you used). See an example here:
https://github.com/cleanlab/examples/blob/fbb88a4055cb7f942520dc59409551a559de71dc/huggingface_keras_imdb/requirements.txt

If possible, try to use versions that are compatible with our main requirements.txt file for the whole examples/ repo:
https://github.com/cleanlab/examples/blob/fbb88a4055cb7f942520dc59409551a559de71dc/requirements.txt

aditya1503 replied:
Done: https://github.com/cleanlab/examples/pull/27/commits/28f62b8528fdc7aec2fd87bca2d5da8e142b7b1e

aditya1503 committed:
requested changes

aditya1503 committed:
more changes

aditya1503 committed:
Merge branch 'multilabel_tutorial' of github.com:aditya1503/examples into multilabel_tutorial

aditya1503 committed:
removing is_multilabel

aditya1503 committed:
verbose + cleanup

aditya1503 committed:
cleanup

jwmueller committed:
wrote text for multilabel classification example notebook

jwmueller committed:
docstring for multilabel network class

jwmueller committed:
clarify dropping rare class-combos

jwmueller committed:
move dropping clarification up

jwmueller committed:
no dash in celebA

jwmueller committed:
one or more clarification

jwmueller committed:
remove installation code cell

aditya1503 committed:
more requirements

jwmueller commented on multilabel_classification/requirements.txt:
```diff
@@ -0,0 +1,7 @@
+torch==1.12.1+cu113
+timm==0.6.7
+pandas==1.5.0
+torchvision==0.13.1+cu102
```
```suggestion
```

jwmueller replied:
I don't see where you need torchvision?

aditya1503 replied:
img = torchvision.io.read_image(path).swapaxes(0, 2).swapaxes(0, 1)

jpeg is not read by PIL/ other libraries, I could add just torch vision without CUDA, 
I'll also see if there are other libraries that can easily display jpeg images

aditya1503 replied:
This was me trying to use torch vision to download CelebA dataset, I'll keep this resolve open, once I figure this out I'll close it

aditya1503 committed:
moving tutorial code

aditya1503 committed:
download fix

aditya1503 committed:
removing hidden cells

aditya1503 committed:
Update multilabel_classification/requirements.txt

Co-authored-by: Jonas Mueller <1390638+jwmueller@users.noreply.github.com>

aditya1503 committed:
Merge branch 'cleanlab:master' into multilabel_tutorial

aditya1503 committed:
executed cells

aditya1503 committed:
Merge branch 'multilabel_tutorial' of github.com:aditya1503/examples into multilabel_tutorial

aditya1503 committed:
cleaning

aditya1503 committed:
execution count

jwmueller committed:
rename notebook

jwmueller committed:
text rewrites and formatting

jwmueller committed:
finalized text for example notebooks

jwmueller reviewed (approved):
Made another final pass of text/code edits, so please review these and give me thumbs up before I merge this.

TODOs for follow-up PR after this is merged (and the official links to these notebooks are available):
- [x] make sure the links in these notebooks to one another are correct
- [x] make sure the link to these notebooks in the tutorial notebook is correct
- [x] add notebooks + short description to table of contents in main readme of this repo

aditya1503 commented:
> Made another final pass of text/code edits, so please review these and give me thumbs up before I merge this.
> 
> TODOs for follow-up PR after this is merged (and the official links to these notebooks are available):
> 
> * [ ]  make sure the links in these notebooks to one another are correct
> * [ ]  make sure the link to these notebooks in the tutorial notebook is correct
> * [ ]  add notebooks + short description to table of contents in main readme of this repo

OK

jwmueller committed:
clarification re empty list label, thanks @elisno

aditya1503 committed:
url change from kaggle to official website

aditya1503 committed:
remove kaggle from training notebook;

aditya1503 committed:
readme changes

aditya1503 committed:
readme url fix

aditya1503 committed:
readme simpler

aditya1503 committed:
readme find label errors

aditya1503 commented:
> Made another final pass of text/code edits, so please review these and give me thumbs up before I merge this.
> 
> TODOs for follow-up PR after this is merged (and the official links to these notebooks are available):
> 
> * [x]  make sure the links in these notebooks to one another are correct
> * [x]  make sure the link to these notebooks in the tutorial notebook is correct
> * [x]  add notebooks + short description to table of contents in main readme of this repo

LGTM! 

minor change: [Kaggle](https://www.kaggle.com/datasets/jessicali9530/celeba-dataset) link is unofficial, so I replaced them with the original dataset [link](https://mmlab.ie.cuhk.edu.hk/projects/CelebA.html)
I've added the notebooks to README.md
